// @vetclinic/shared-kernel/dist/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestValidator = exports.BaseController = exports.ErrorGuards = exports.ErrorHandler = exports.AuthorizationError = exports.DuplicateError = exports.NotFoundError = exports.ValidationError = exports.AppError = void 0;
var AppError_1 = require("./AppError");
Object.defineProperty(exports, "AppError", { enumerable: true, get: function () { return AppError_1.AppError; } });
var ValidationError_1 = require("./ValidationError");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return ValidationError_1.ValidationError; } });
var NotFoundError_1 = require("./NotFoundError");
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return NotFoundError_1.NotFoundError; } });
var DuplicateError_1 = require("./DuplicateError");
Object.defineProperty(exports, "DuplicateError", { enumerable: true, get: function () { return DuplicateError_1.DuplicateError; } });
var AuthorizationError_1 = require("./AuthorizationError");
Object.defineProperty(exports, "AuthorizationError", { enumerable: true, get: function () { return AuthorizationError_1.AuthorizationError; } });
var ErrorHandler_1 = require("./ErrorHandler");
Object.defineProperty(exports, "ErrorHandler", { enumerable: true, get: function () { return ErrorHandler_1.ErrorHandler; } });
var ErrorGuards_1 = require("./ErrorGuards");
Object.defineProperty(exports, "ErrorGuards", { enumerable: true, get: function () { return ErrorGuards_1.ErrorGuards; } });
var BaseController_1 = require("./BaseController");
Object.defineProperty(exports, "BaseController", { enumerable: true, get: function () { return BaseController_1.BaseController; } });
var RequestValidator_1 = require("./RequestValidator");
Object.defineProperty(exports, "RequestValidator", { enumerable: true, get: function () { return RequestValidator_1.RequestValidator; } });
